File: leetcode/graph/dfs/0417_pacific_atlantic_water_flow.py
# LC 417. Pacific Atlantic Water Flow | Medium
class Solution:
    def pacificAtlantic(self, heights):
        if not heights: return []
        m, n = len(heights), len(heights[0]); pac, atl = set(), set()
        def dfs(i, j, visited, prev_h):
            if i < 0 or i >= m or j < 0 or j >= n: return
            if (i,j) in visited or heights[i][j] < prev_h: return
            visited.add((i,j))
            for di, dj in [(0,1),(0,-1),(1,0),(-1,0)]:
                dfs(i+di, j+dj, visited, heights[i][j])
        for j in range(n): dfs(0, j, pac, 0); dfs(m-1, j, atl, 0)
        for i in range(m): dfs(i, 0, pac, 0); dfs(i, n-1, atl, 0)
        return list(pac & atl)

File: leetcode/graph/dfs/0417_pacific_atlantic_water_flow.cpp
// LC 417. Pacific Atlantic Water Flow | Medium
#include <vector>
using namespace std;

class Solution {
    int m, n;
    void dfs(vector<vector<int>>& h, vector<vector<bool>>& vis, int i, int j) {
        vis[i][j] = true;
        int d[] = {0,1,0,-1,0};
        for (int x = 0; x < 4; x++) {
            int ni = i+d[x], nj = j+d[x+1];
            if (ni >= 0 && ni < m && nj >= 0 && nj < n && !vis[ni][nj] && h[ni][nj] >= h[i][j])
                dfs(h, vis, ni, nj);
        }
    }
public:
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {
        m = heights.size(); n = heights[0].size();
        vector<vector<bool>> pac(m, vector<bool>(n)), atl(m, vector<bool>(n));
        for (int j = 0; j < n; j++) { dfs(heights, pac, 0, j); dfs(heights, atl, m-1, j); }
        for (int i = 0; i < m; i++) { dfs(heights, pac, i, 0); dfs(heights, atl, i, n-1); }
        vector<vector<int>> res;
        for (int i = 0; i < m; i++) for (int j = 0; j < n; j++)
            if (pac[i][j] && atl[i][j]) res.push_back({i, j});
        return res;
    }
};
